### DISCUSSION TITLE
CORS blocking fetching of websites

### DISCUSSION THREAD

**Discussion by caleb-at-pieces** [jupyterlab/jupyterlab#15932] [Q&A]
Hello,

I develop a JupyterLab extension for pieces.app, and I have recently added a feature that will allow users to utilize websites as context within our copilot chat.

However, I am getting blocked by CORS while trying to fetch webpages, does anyone know a good workaround for this that does not require me to run my own server?

<img width="1061" alt="image" src="https://github.com/jupyterlab/jupyterlab/assets/133268200/918c905d-ac52-44e6-a3ea-d9b420d1951f">

Thanks

**Comment by Labeeb2339** [comment]
CORS is enforced by the browser, not by JupyterLab, so a frontend extension cannot make an arbitrary cross-origin page readable unless that page opts in with the appropriate `Access-Control-Allow-Origin` (and, when applicable, preflight) headers.

The usual options are:

- Have the target site enable CORS for the exact extension origin, if you control it.
- Proxy the request through a server you control, such as a small Jupyter Server extension or your own backend. The extension then fetches a same-origin endpoint, and the server performs the outbound request.
- If the target site does not allow either direct CORS or a server-side proxy, there is no reliable client-only workaround. `mode: "no-cors"` returns an opaque response that JavaScript still cannot read, and JSONP only applies to APIs that explicitly support it.

To debug this, inspect the browser Network panel for the request and any `OPTIONS` preflight. The response headers will show whether the target allows the origin, methods, and headers your extension is sending. Also compare the origin of the JupyterLab page (often `http://localhost:8888`) with the URL being fetched.

**Comment by rootkiller6788** [comment]
One thing that reframes the whole problem: you already have a server — the Jupyter Server
  your extension is hosted on. You don't need to stand up your own; you need a server extension
  to do the fetch.

  CORS is enforced by the browser, so a frontend-only extension can't read an arbitrary
  cross-origin page unless that site sends Access-Control-Allow-Origin for your origin. There's
  no client-side escape hatch — no-cors returns an opaque response you still can't read, and
  JSONP only works where the target site implements it.

  The idiomatic fix is to fetch server-side. Add a server extension exposing a small handler on
  the same Jupyter Server, something like:

  class FetchHandler(web.RequestHandler):
      async def get(self):
          url = self.get_argument("url")
          async with httpx.AsyncClient() as client:
              r = await client.get(url)
          self.set_header("Content-Type", "text/html")
          self.write(r.text)

  register it at a route (say /pieces/fetch?url=...), and have your frontend call that
  same-origin endpoint instead of the target directly. Server-to-server requests aren't subject
  to CORS, so any page works. That's the standard shape for this: your extension ships the
  frontend TypeScript you already have, plus a small Python server part.

  jupyter-server-proxy is the general-purpose version of this if you'd rather not write the
  handler yourself, though it's aimed at proxying to a known service more than arbitrary URLs.

  One caveat: server-side fetch gets you past CORS but not past a site that outright refuses
  non-browser requests or needs the user's cookies. If those pages also 403 from Python, that's
  a separate problem.
